src/core/server/http/router/router.ts

```diff
@@ -31,116 +31,175 @@ interface RouterRoute {
- * ```
- *
+ * Registers route handlers for specified resource path and method.
  * @public
- * */
-export class Router {
-  private routes: Array<Readonly<RouterRoute>> = [];
-
-  constructor(readonly path: string) {}
+ */
+export interface IRouter {
+  /**
+   * Resulted path
+   */
+  routerPath: string;
 
   /**
    * Register a route handler for `GET` request.
    * @param route {@link RouteConfig} - a route configuration.
    * @param handler {@link RequestHandler} - a function to call to respond to an incoming request
    */
+  get: <P extends ObjectType, Q extends ObjectType, B extends ObjectType>(
+    route: RouteConfig<P, Q, B>,
+    handler: RequestHandler<P, Q, B>
+  ) => void;
+
+  /**
+   * Register a route handler for `POST` request.
+   * @param route {@link RouteConfig} - a route configuration.
+   * @param handler {@link RequestHandler} - a function to call to respond to an incoming request
+   */
+  post: <P extends ObjectType, Q extends ObjectType, B extends ObjectType>(
+    route: RouteConfig<P, Q, B>,
+    handler: RequestHandler<P, Q, B>
+  ) => void;
+
+  /**
+   * Register a route handler for `PUT` request.
+   * @param route {@link RouteConfig} - a route configuration.
+   * @param handler {@link RequestHandler} - a function to call to respond to an incoming request
+   */
+  put: <P extends ObjectType, Q extends ObjectType, B extends ObjectType>(
+    route: RouteConfig<P, Q, B>,
+    handler: RequestHandler<P, Q, B>
+  ) => void;
+
+  /**
+   * Register a route handler for `DELETE` request.
+   * @param route {@link RouteConfig} - a route configuration.
+   * @param handler {@link RequestHandler} - a function to call to respond to an incoming request
+   */
+  delete: <P extends ObjectType, Q extends ObjectType, B extends ObjectType>(
+    route: RouteConfig<P, Q, B>,
+    handler: RequestHandler<P, Q, B>
+  ) => void;
+
+  /**
+   * Returns all routes registered with the this router.
+   * @returns List of registered routes.
+   * @internal
+   */
+  getRoutes: () => RouterRoute[];
+}
+
+export type ContextEnhancer<P extends ObjectType, Q extends ObjectType, B extends ObjectType> = (
+  handler: RequestHandler<P, Q, B>
+) => RequestHandlerEnhanced<P, Q, B>;
+
+function getRouteFullPath(routerPath: string, routePath: string) {
+  // If router's path ends with slash and route's path starts with slash,
+  // we should omit one of them to have a valid concatenated path.
+  const routePathStartIndex = routerPath.endsWith('/') && routePath.startsWith('/') ? 1 : 0;
+  return `${routerPath}${routePath.slice(routePathStartIndex)}`;
+}
+
+/**
+ * @internal
+ */
+export class Router implements IRouter {
+  public routes: Array<Readonly<RouterRoute>> = [];
+
+  constructor(
+    readonly routerPath: string,
+    private readonly log: Logger,
+    private readonly enhanceWithContext: ContextEnhancer<any, any, any> = fn => fn.bind(null, {})
+  ) {}
+
```

Comment:
Now that all the docs are moved to the interface, you might be able to simplify this implementation quite a bit by generating these functions in the constructor:

```ts
class Router implements IRouter {

  constructor() {
    readonly routerPath: string,
    private readonly log: Logger,
    private readonly enhanceWithContext: ContextEnhancer<any, any, any> = fn => fn.bind(null, {})
  {
    // Kept local to constructor to hide from consumers
    const buildMethod = (method: string) => <P extends ObjectType, Q extends ObjectType, B extends ObjectType>(
      route: RouteConfig<P, Q, B>,
      handler: RequestHandler<P, Q, B>
    ) {
      // do the repetitive work here
    };

    this.get = buildMethod('get');
    this.post = buildMethod('post');
  }
}
```

This would also make it much simpler to add methods in the future.